aws/data_source_aws_caller_identity.go

```diff
@@ -48,6 +56,38 @@ func dataSourceAwsCallerIdentityRead(d *schema.ResourceData, meta interface{}) e
 	d.Set("account_id", res.Account)
 	d.Set("arn", res.Arn)
 	d.Set("user_id", res.UserId)
+	d.Set("source_arn", sourceARN(aws.StringValue(res.Arn)))
 
 	return nil
 }
+
+// sourceARN returns the same string passed in unless it appears to be an assumed role ARN.
+// In that case, it attempts to return the source role ARN associated with an assumed role ARN.
+func sourceARN(rawARN string) string {
+	result := rawARN
+
+	if strings.Contains(result, ":assumed-role/") && strings.Contains(result, ":sts:") {
+		parsedARN, err := arn.Parse(result)
+
+		if err != nil {
+			return result
+		}
+
+		if parsedARN.Service != "sts" {
+			// not an assumed role
+			return result
+		}
+
+		re := regexp.MustCompile(`^assumed-role/`)
+		parsedARN.Resource = re.ReplaceAllString(parsedARN.Resource, "role/")
```

Comment:
This will return the incorrect sourceArn for roles with paths. The sts assumed-role ARN doesn't include the path of the role. For instance, if you have IAM role ARN `arn:aws:iam::{accountId}:role/{path}/{roleName}` and you assume that role , the assumed-role ARN will be `arn:aws:sts::{accountId}:assumed-role/{roleName}/{sessionName}`. To derive the sourceARN, you would need to call getRole after parsing the role name from the sts ARN. The problem is then, the caller identity data source will then require the `iam:GetRole` permission, which could theoretically break workflows.

Reply:
😞 absolutely right